Add pavex_session_redis lib for Redis-based session management

joehasson opened the pull request:
Hey, I'm working on #466 following the patterns used in the postgres-based session store. I have a working implementation and there's still some work to do polish this (mainly tests + documentation) but I'd like to get some pointers around architectural patterns beforehand, any guidance is much appreciated.

1. Exposing redis-rs as a dependency: this new lib defines a `RedisSessionStore` constructor with a `new` method accepting a `redis::aio::ConnectionManager`. This is analogous to the situation with the `PostgresSessionStore` constructor in the postgres implementation which accepts a `sqlx::PgPool`. This means that the user of these libraries is forced to construct these types and hence to depend directly on `redis-rs` / `sqlx`. I'm wondering whether it would be better, at least in the redis case, if the dependency was instead abstracted away? We could instead just require the user to supply connection strings rather than pre-constructed clients.
2. Injecting configuration:  it would be nice to let users configure how redis keys are generated, [like actix-session does](https://docs.rs/actix-session/latest/actix_session/storage/struct.RedisSessionStoreBuilder.html#method.cache_keygen) , so that they can implement namespacing and use existing redis stores to store their session data. It's not clear to me how to do that with the current approach, however, since dependency injection happens at compile time. 

One way to address both of these points might be to have the `RedisSessionStore::new` function instead accept a struct of the following type, which we would require the user to instantiate:
```
pub struct RedisSessionStoreConfig {
    pub connection_string: String,
    pub key_generator: Option<Box<dyn Fn(&str) -> String>>,
}
```
Does this seem reasonable? (If we definitely want to expose redis-rs as a dependency this struct could store a `ConnectionManager` instead of a connection string). 

3. Testing: as mentioned in https://github.com/LukeMathWalker/pavex/pull/497, if you have any particular thoughts on how to test this stuff I'm keen to hear them, otherwise I am happy to crack on.

Many thanks for any input on the above questions and any other comments on the PR.

LukeMathWalker commented on libs/pavex_session_redis/src/lib.rs:
```diff
@@ -0,0 +1,327 @@
+        let ttl = match ttl_reply {
+            Value::Int(s) if s >= 0 => std::time::Duration::from_secs(s as u64),
+            Value::Int(-1) => {
+                panic!("Fatal session management error: no TTL set for this session.")
```
We shouldn't panic.

LukeMathWalker commented on libs/pavex_session_redis/src/lib.rs:
```diff
@@ -0,0 +1,327 @@
+                .map_err(LoadError::DeserializationError)?,
+            _ => {
+                return Err(LoadError::Other(anyhow::anyhow!(
+                    "Redis GET replied {:?}. Expected bulk string.",
```
This _may_ leak the stored data, which might contain sensitive information.
We should limit ourselves to stating which value type we got, rather than logging the value itself.

LukeMathWalker commented on libs/pavex_session_redis/src/lib.rs:
```diff
@@ -0,0 +1,327 @@
+            Value::Int(-2) => return Ok(None),
+            _ => {
+                return Err(LoadError::Other(anyhow::anyhow!(
+                    "Redis TTL returned {:?}. Expected integer >= -2",
```
Same as below, regarding privacy.

joehasson commented:
Hi @LukeMathWalker, thanks very much for the feedback above. I believe these commits address everything we've discussed. Worth calling out on this point regarding the dedicated configuration type
> In turn, this forces us to decide if a fully-general key_generator is needed or a simpler key_prefix: Option<String> is enough.

Given that the only use case I had in mind  here was namespacing, a fully general key_generator seems on reflection like overkill and an unnecessary footgun, so I went ahead and implemented the `Option<String>` alternative. 

For the tests, I have mostly copy-pasted from @oliverbarnes' work on https://github.com/LukeMathWalker/pavex/pull/497 since the relevant tests carry over with the appropriate changes to the `create_test_store` helper. It strikes me that it would be helpful to create a single test suite which was generic over the `SessionStorageBackend` trait as more backends are implemented, to prevent this kind of duplication.

Happy to squash these commits if you'd prefer a cleaner history when its time to merge.

oliverbarnes commented:
Hey @joehasson, just a heads-up that I've made some changes to the tests today, based on Luca's feedback. Namely I've moved them to a separate file, and cleaned up a couple of superfluous assertions.

I'm curious about your idea to abstract the test suite, specially given I'm now implementing a very similar [suite for MySQL](https://github.com/LukeMathWalker/pavex/pull/499/files#diff-ee6ec41746f54040a44b69252daee0aa52ca62da4b21aece32e2c7aab6e8e224) on #499.

...though I have to say, coming from Rails/Rspec where sometimes people went a little overboard with test abstractions, that these days I have a lot more tolerance for test duplication =D 

(tangentially: "footgun" is a great word, thanks - first time I see it)

joehasson commented:
Point well taken from both of you regarding test deduplication - seems more effort than it's worth on reflection.
So I have brought my branch up to date in light of subsequent discussion on #497. To be completely explicit:

- Refactored tests in line with comments on #497 addressed in a0faea84d11e8b0b4e6773938927bfe8b20e2240
- Added unhappy path tests, broadly in line with those in c6bf07b10d6263261c02bd4f1539cfa28e38f568. Code coverage is now in the high 80s as measured by cargo llvm-cov. I have omitted a counterpart to the `test_database_unavailable_error` test since Redis ConnectionManager doesn't have an equivalent to SQLite's pool.close() method, and simulating network failures would require external dependencies or system-level changes that seem like overkill for a unit test of a relatively simple code-path.
- Moved tests to their own folder in line with 316634aaa6da36a7fd337823f8fda67016d27b6a

I'll convert this from draft once I have updated and added documentation as mentioned before, unless there's any further comments. Thanks very much!

LukeMathWalker commented on libs/pavex_session_redis/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+#[derive(Clone, Debug, Default, serde::Deserialize)]
+/// Configuration options for the Redis session store.
+pub struct RedisSessionStoreConfig {
+    /// Optional namespace prefix for Redis keys. When set, all session keys will be prefixed with this value.
+    #[serde(default)]
+    pub namespace: Option<String>,
```
Let's include an example, showing the structure of a namespaced key (i.e. that `:` is used as separator).

LukeMathWalker commented on libs/pavex_session_redis/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    }
+}
+
+pub struct RedisSessionKit {
```
There's no longer a need for kits. It's enough to add a couple of attributes to get thing working with the new DI system.
You can use either pavex.dev/docs or the MySQL PR as a reference on this matter.
